Suggested style changes to Multiselect, MulticolumnList

annamelnyk opened the pull request:
# Description
Suggested style changes to improve components layout:

- wrapping of long words for Multiselect and MulticolumnList cells
- make filtered items on full width in Multiselect
- make MulticolumnList cells border single height 

# Screenshots
**Before**
![image](https://user-images.githubusercontent.com/55694637/78126666-886ceb80-741b-11ea-8694-e19ec9c00ba9.png)
![image](https://user-images.githubusercontent.com/55694637/78126715-9cb0e880-741b-11ea-9e9f-d00e965fb5ff.png)

**After**
![image](https://user-images.githubusercontent.com/55694637/78142236-58cadd00-7435-11ea-8a33-045ce530da68.png)

![image](https://user-images.githubusercontent.com/55694637/78126827-cd911d80-741b-11ea-8bfe-d56f5c14a7e5.png)

rasmuswoelk commented on lib/MultiColumnList/MCLRenderer.css:
```diff
@@ -163,16 +163,28 @@
+    background-color: var(--color-border-p2);
+  }
 
   &:last-child {
```
The border should be removed on the last child.

![image](https://user-images.githubusercontent.com/640976/78131838-fdd8bc00-741b-11ea-998b-ed9cd5e54f72.png)

rasmuswoelk commented on lib/MultiColumnList/MCLRenderer.css:
```diff
@@ -163,16 +163,28 @@
+    position: absolute;
+    right: 0;
+    top: 5px;
+    bottom: 5px;
```
We have a variable that has this value (5px): `--gutter-static-one-third`

rasmuswoelk commented on lib/MultiSelection/MultiSelect.css:
```diff
@@ -159,6 +159,7 @@
   display: inline-block;
   border: 1px solid var(--color-border);
   background-color: #efefef;
+  width: 100%;
```
This is not ideal because we won't be taking advantage of the available space in most cases:

![image](https://user-images.githubusercontent.com/640976/78379466-b2b4d980-75d2-11ea-86d7-ea272b56e629.png)

rasmuswoelk replied:
I think the word-break alone is enough:

![image](https://user-images.githubusercontent.com/640976/78379631-fb6c9280-75d2-11ea-8a9f-1c81557a40c4.png)

rasmuswoelk commented on lib/MultiColumnList/MCLRenderer.css:
```diff
@@ -163,19 +163,34 @@
+  word-break: break-word;
+  position: relative;
+
+  &::after {
+    content: '';
+    position: absolute;
+    right: 0;
+    top: var(--gutter-static-one-third);
+    bottom: var(--gutter-static-one-third);
+    width: 1px;
+    background-color: var(--color-border-p2);
+  }
```
Taking another look at this made me realize that this change will be applied to all MultiColumnList's. So while it will work fine for EditableList's specifically, it won't play nice with basic MCL's:

![image](https://user-images.githubusercontent.com/640976/78379835-48506900-75d3-11ea-81fd-8d8efff6e61b.png)

These borders should go from top to bottom without a gap.

rasmuswoelk replied:
Would it maybe be possible to make these changes for the EditableList specifically instead?

JohnC-80 replied:
Could we A) break out borders/modifications into their own classes, keeping the more functional cell class intact and b) provide a prop that might allow you to set a custom class in addition to the existing classes? It seems like it would add the flexibility you need for the use-case of EditableList, @annamelnyk

rasmuswoelk commented on lib/MultiColumnList/MCLRenderer.css:
```diff
@@ -186,7 +201,6 @@
 .mclScrollable {
   flex-grow: 1;
   flex-shrink: 0;
-  overflow: auto;
```
Why was this removed?

annamelnyk replied:
It shoud not be removed, I've tested one more time and moved it back

annamelnyk commented:
@rasmuswoelk, @JohnC-80 I fixed comment and added `borderWithGaps` prop, setted to `false` by default, and to `true` for Editable components. Please take a look for [this](https://github.com/folio-org/stripes-smart-components/pull/765) also, it should be merged with this PR.

rasmuswoelk commented:
> @rasmuswoelk, @JohnC-80 I fixed comment and added `borderWithGaps` prop, setted to `false` by default, and to `true` for Editable components. Please take a look for [this](https://github.com/folio-org/stripes-smart-components/pull/765) also, it should be merged with this PR.

Hi @annamelnyk 

The `MultiColumnList` is already pretty bloated with a lot of different props so I think we need a solution that is a bit more versatile and can be re-used for other purposes as well.

 @JohnC-80 and I discussed a possible approach for this. We ended up with a prop that could be used to apply additional CSS classes for columns in an MCL. This prop could then be used to apply the custom CSS for the EditableList in stripes-smart-components since it's the only place currently (afaik) that we need borders with gaps. E.g.:

```js
className = {`${css.mclCell} ${showOverflow} ${props.getCellClasses ? props.getCellClasses(css.mclCellStyle) : css.mclCellStyle}`}
```

So the `getCellClases`-prop would receive the default style for the column so that it could either be extended:

```js
getCellClasses={defaultClass => `${defaultClass} ${css.myCustomClass}`
```

... or completely overwritten:
```js
getCellClasses={() => css.myCustomClass}
```

The idea is to separate the functional classes from the visual style classes so `css.mclCell` would still contain the functional CSS whereas e.g. a `css.mclCellStyle`-class would contain the default style for the cell – in this case, it would be the border, min-height, and line-height (if I remember correctly).

I hope this makes sense. Otherwise, feel free to write to me on Slack or here in the PR. I'm sorry for the delayed response to this and the additional work needed to achieve such a little change.. Almost nothing is simple in a Stripes project :-D

And @JohnC-80 if you have any additions or corrections then feel free to comment too 👍

JohnC-80 commented:
@rasmuswoelk 's  comment is exactly what we need for this, @annamelnyk  - MCL already has many props - if we merge this as-is and eventually need something to affect padding, background-color, whathaveyou,  we would need another prop and another prop and another - so allowing the instance level to supply a function for setting the aesthetic class is most ideal.

id-jenkins commented:
<html><body><pre>
yarn run v1.22.4
$ eslint ./ && stylelint "lib/**/*.css"

/home/jenkins/workspace/o-org_stripes-components_PR-1230/project/lib/MultiColumnList/MCLRenderer.js
  1141:1  error  Trailing spaces not allowed  no-trailing-spaces

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

annamelnyk commented:
@rasmuswoelk @JohnC-80 please take a look to changes, I am not sure that current implementation is what you expected. I was doubt if I should pass to Editable list prop `getCellClasses` as function, but I don't see how can it be done if all styles are in `MCLRenderer.css`. [Changes in EditableList](https://github.com/folio-org/stripes-smart-components/pull/765)

id-jenkins commented:
<html><body><pre>
yarn run v1.22.4
$ eslint ./ && stylelint "lib/**/*.css"

/home/jenkins/workspace/o-org_stripes-components_PR-1230/project/lib/MultiColumnList/MCLRenderer.js
  1141:1  error  Trailing spaces not allowed  no-trailing-spaces

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>

id-jenkins commented:
<html><body><pre>
yarn run v1.22.4
$ eslint ./ && stylelint "lib/**/*.css"

/home/jenkins/workspace/o-org_stripes-components_PR-1230/project/lib/MultiColumnList/MCLRenderer.js
  1141:1  error  Trailing spaces not allowed  no-trailing-spaces

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
</pre><body></html>
